test/extensions/filters/http/cache/cache_filter_test.cc

```diff
@@ -5,122 +5,253 @@
+                      /*metadata=*/_, /*end_stream=*/false));
+    for (const absl::string_view& chunk : body_chunks) {
+      bool end_stream = (&chunk == &body_chunks.back());
+      EXPECT_CALL(*insert_context, insertBody(bufferContains(chunk), _, end_stream))
+          .WillOnce(InvokeArgument<1>(true));
+    }
+    EXPECT_CALL(*insert_context, onDestroy());
 
-    // The filter should stop decoding iteration when decodeHeaders is called as a cache lookup is
-    // in progress.
-    EXPECT_EQ(filter->decodeHeaders(request_headers_, true),
-              Http::FilterHeadersStatus::StopAllIterationAndWatermark);
+    return std::move(insert_context);
+  }
 
-    // The filter should not continue decoding when the cache lookup result is ready, as the
-    // expected result is a hit.
-    EXPECT_CALL(decoder_callbacks_, continueDecoding).Times(0);
+  // Passes request_headers_ into the filter's decodeHeaders(), and runs any
+  // subsequent callbacks posted to the dispatcher. Returns the return value of
+  // the encodeHeaders() call.
+  Envoy::Http::FilterHeadersStatus decodeHeadersAndRunDispatcher(bool end_stream = true) {
+    Envoy::Http::FilterHeadersStatus decode_headers_result =
+        filter_->decodeHeaders(request_headers_, end_stream);
+    dispatcher_->run(Envoy::Event::Dispatcher::RunType::Block);
+    return decode_headers_result;
+  }
 
-    // The cache lookup callback should be posted to the dispatcher.
-    // Run events on the dispatcher so that the callback is invoked.
-    // The posted lookup callback will cause another callback to be posted (when getBody() is
-    // called) which should also be invoked.
-    dispatcher_->run(Event::Dispatcher::RunType::Block);
+  void waitBeforeLookup() { time_source_.advanceTimeWait(delay_); }
 
-    ::testing::Mock::VerifyAndClearExpectations(&decoder_callbacks_);
+  testing::Matcher<const Envoy::Http::ResponseHeaderMap&>
+  isSupersetOfCacheableHeaders(const Envoy::Http::TestResponseHeaderMapImpl& headers) const {
+    Envoy::Http::TestResponseHeaderMapImpl cacheable_response_headers = headers;
+    for (const std::string& header : uncached_headers_) {
+      cacheable_response_headers.remove(Envoy::Http::LowerCaseString(header));
+    }
+    return testing::AllOf(IsSupersetOfHeaders(cacheable_response_headers),
+                          doesNotHaveHeaders(uncached_headers_));
   }
 
-  void waitBeforeSecondRequest() { time_source_.advanceTimeWait(delay_); }
+  testing::Matcher<const Envoy::Http::ResponseHeaderMap&>
+  isSupersetOfCacheableResponseHeaders() const {
+    return isSupersetOfCacheableHeaders(response_headers_);
+  }
+
+  testing::Matcher<Envoy::Http::ResponseHeaderMap&> isSupersetOfResponseHeadersWithAge() {
+    return testing::AllOf(IsSupersetOfHeaders(response_headers_),
+                          HeaderHasValueRef(Http::CustomHeaders::get().Age, age_));
+  }
 
-  SimpleHttpCache simple_cache_;
+  std::shared_ptr<StrictMock<MockHttpCache>> cache_;
```

Comment:
Please document why this test needs the additional complexity of using a mock and having to respecify its behavior.

Reply:
Added a test fixture comment. This is most important for flow control tests in a future change - I can imagine splitting the test fixture to let the tests that don't need as fine-grained control use a fake, but for now this is at least an improvement over specifying the fake's behavior via extra copies of CacheFilter.

Reply:
It's hard to add those tests before the CacheFilter implements its flow control logic: right now we can't test e.g. downstream backpressure while serving from cache because the whole response gets buffered, so there aren't any signals from the connection manager that we can interleave with cache body fragments.

I take your point about it being hard to trust that complex mocks are doing the right thing. My thinking is that when it comes to setting up the environment and reading results, it's even worse to trust that the code under test is doing the right thing. So in the existing state of the test, we test CacheFilter inserts by using CacheFilter lookups to verify the state of the cache, and we test lookups by using inserts to set up the cache, but we can never verify the behavior of one of those operations independently of the other. So there's a class of bugs that affect both insert and lookup that the existing test setup can never catch.

Reply:
Testing inserts and lookups independently is not a goal, because they have practically no independently defined behavior. The (simplified) spec is that if you insert something then look it up, you get the response you inserted (with certain modifications). We don't use CacheFilter lookups to verify the state of the cache--we use CacheFilter lookups to verify that the observable behavior of a particular sequence of operations is correct. If some "bug" doesn't change the observable behavior, then it isn't really a bug.

There are times when it is helpful to write tests that are less faithful to deployed reality, and testing a slow HttpCache is probably such a time. However, this PR changes all of the tests in this file to using mocks, including those that are better off using the real thing. Perhaps CacheFilterTest should have RealCacheFilterTest and MockCacheFilterTest subclasses, or should use a simple mock that defaults to letting SimpleHttpCache handle everything, with ON_CALLs only appearing in specific tests?

Reply:
In the past I've done some amount of insert/lookup testing against stats held in the cache object. That allows some modest testing of insert/lookup in isolation. E.g. you could verify that inserts fill the cache with data without doing a lookup. You could verify that lookups fail due to expiry. This doesn't require mocks.

@mbehr maybe a spec of what your testing plan is would make it easier to understand the future value of the mocks.